### ISSUE TITLE
Migrate `RouterModule.forChild()` calls with `ng g @angular/core:standalone`

### ISSUE THREAD

**Issue by cexbrayat** [angular/angular#48950]
### Which @angular/* package(s) are relevant/related to the feature request?

core

### Description

When migrating an application with lazy-loaded modules, a fairly recurrent pattern is encountered:

app.routes.ts
```ts
export const routes = [
  {
     path: 'admin',
     loadChildren: () =>
      import('./admin/admin.module').then(m => m.AdminModule)
   }
];
```

admin.module.ts
```ts
imports: [RouterModule.forChild(adminRoutes)]
```

admin.routes.ts
```ts
export const adminRoutes = [
  { path: '', component: AdminComponent }
]
```

Currently, the migration does not touch `RouterModule.forChild()` calls. This then prevents the `prune-ng-module` step to remove the module.

### Proposed solution

It would be great to migrate `app.routes.ts` to:

```ts
export const routes = [
  {
     path: 'admin',
     loadChildren: () =>
      import('./admin/admin.routes').then(m => m.adminRoutes)
   }
];
```

and remove `RouterModule.forChild(adminRoutes)` form `admin.module.ts`

### Alternatives considered

The migration can be done manually.

**Comment by atscott** [comment]
This transformation can potentially be done under the following rules:

1. Every component in the child routes has been converted to `standalone`
2. `AdminModule` does not have any providers or initialization logic (i.e. in its constructor). If it does have providers, we could probably still make the transformation, but the resulting route would have to have the additional property `providers: importProvidersFrom(AdminModule)`.

**Comment by elvisbegovic** [comment]
Great idea, I was going to create this request and I found yours (even if mine is less specific to routes).

1) 
Thanks to @crisbeto , on our relatively medium project, execution of [MigrationMode.toStandalone](https://github.com/angular/angular/commit/a154db8a81cbdfed8c3d0db1e2a5bf43aa3e0bbf) works like a charm
**1532 components UPDATED !** 

![image](https://user-images.githubusercontent.com/2789203/229748505-8f5c0bf7-7e40-4eff-aa33-55a45188f41b.png)

2) 
Then I tried [MigrationMode.pruneModules](https://github.com/angular/angular/commit/e7318fc758d9e64d1a7f60a2c7071a769b73e7d8) **only 1 module is deleted**. 

![image](https://user-images.githubusercontent.com/2789203/229750311-08d39089-38c8-42e5-b1b4-11f0edf4815d.png)

3) 
We almost have all ngmodules with : 
```
const routes: Routes = [{path: '', component: ApparComponent}] as Routes;

@NgModule({
  imports: [
    ShareModule,
    RouterModule.forChild(routes)
  ]
})
export class ApparModule {}
```

4)
It is out of scope to me, I just tried to add some logs in [canRemoveClass ](https://github.com/angular/angular/commit/e7318fc758d9e64d1a7f60a2c7071a769b73e7d8#diff-14301c4864bcadfcdc0f064dae26121a553e46e286cf7aca984fd4c76d20cd3eR201-R249) like : 

![image](https://user-images.githubusercontent.com/2789203/229769351-f6e35caa-7404-40ae-80c9-ed33959103c8.png)

When I manually delete `RouterModule.forChild(routes)` from `ApparModule `, I still run into same behavior due to ngmodule dependancy 'ShareModule' using providers[SomeService].

5)
Question :  it seems step1 is working fully because is easier than step2, but: in adition to curent issue, is there a place to new feat. req. to enhance/migrate automatically ngmodule.providers[SomeService] which principally avoid prune ngmodules. (also there is no // TODO comment in code and no advice or documentation in console) to help developer to know which dep-ng-module is ignoring deletion and what we can do to go ahead. 
What you think ?

**Comment by colas74** [comment]
Same issue here, cannot migrate featured ngmodule due to RouterModule.forChild providers

**Comment by Giggzzz** [comment]
I am facing the same issue to migrate NgModules in an app with 1000+ components. That would be a great update

**Comment by elvisbegovic** [comment]
Dear @atscott will this be highly labelized to help people migrate routes automatically? Or providing migration tooling when possible is now less important nowadays and your advice is to do it manually on +300 routes then how

**Comment by roma2341** [comment]
what if i need to use "RouterModule.forChild" in standalone component ? How I can do it ? Cannot find solution.

**Comment by JoostK** [comment]
> what if I need to use "RouterModule.forChild" in a standalone component? How can I do it manually? Cannot find a solution.

It can't be done, because it's designed not to be possible. `ModuleWithProviders` are "environment injector" (aka "module injectors") providers and cannot be mixed into "component injectors" (aka "element injectors").

**Comment by roma2341** [comment]
@JoostK so now it's not possible to make lazy nested route ? e.g A->B(lazy)->C(lazy).

**Comment by JoostK** [comment]
`Route.providers` does accept the usage of `importProvidersFrom` since it creates an environment injector to create the routed component in.

**Comment by atscott** [comment]
> Dear @atscott will this be highly labelized to help people migrate routes automatically? Or providing migration tooling when possible is now less important nowadays and your advice is to do it manually on +300 routes then how

Migrations are prioritized more highly when APIs are deprecated or removed. Neither `RouterModule.forRoot` nor `RouterModule.forChild` are not deprecated so the migration is not a priority at the moment.

**Comment by roma2341** [comment]
> > Dear @atscott will this be highly labelized to help people migrate routes automatically? Or providing migration tooling when possible is now less important nowadays and your advice is to do it manually on +300 routes then how
> 
> Migrations are prioritized more highly when APIs are deprecated or removed. Neither `RouterModule.forRoot` nor `RouterModule.forChild` are deprecated so the migration is not a priority at the moment.

Yes, but they aren't compatible with standalone components (as I can't use RouterModule.forChild in a standalone component). I don't understand what does importProvidersFrom.. There is no documentation on how to do this stupid migration from module-based components to standalone components where RouterModulel.forChild is being used.

**Comment by anuradhawick** [comment]
@roma2341 just to confirm with you, we cannot create routes as follows right?

`A->B(lazy)->C(lazy)`

I am also having trouble understanding how that could be facilitated. The original code has two nested levels using `RouterModule.forChild(routes)` function. 

There is a chance that we can achieve this by exporting child router array from a nested component back to parent and then to `provideRouter(routes)`. That kind of cluttered things and does not help the business logic separation.

Any suggestions? I am super keen to migrate to standalone.

**Comment by roma2341** [comment]
> @roma2341 just to confirm with you, we cannot create routes as follows right?
> 
> `A->B(lazy)->C(lazy)`
> 
> I am also having trouble understanding how that could be facilitated. The original code has two nested levels using `RouterModule.forChild(routes)` function.
> 
> There is a chance that we can achieve this by exporting child router array from a nested component back to parent and then to `provideRouter(routes)`. That kind of cluttered things and does not help the business logic separation.
> 
> Any suggestions? I am super keen to migrate to standalone.

I stopped understanding how it works =)

**Comment by elvisbegovic** [comment]
Due to first version of angular2+ before aot and slow browser-reload on dev mode  > after aot by default and slow ng serve reload ... or slow angular-cli before webpack , or with webpack older version, or slow ng-cli without hot-reload / no cache / abondonned bazel-compiler ... due to all of this and others reasons, for large code base the recommendations was simple and always : **lazy-loading module / lot lot lot chunks** otherwise app sucks ! 

Result : many and many ngmodule / routermodule is created and this continue nowadays...

If we want follow angular update cadence and if we want drop ngmodule (for what it is created, simplify codesource for new developers) we need "ng update" to be better considered/priorized/allocate more ressrouce and automatically update things it can. 

i know many company suffering of "ng update" not updating things it can programatically/AST

recall, can this be prioritized because it has higher impact on company like recruiting/training new developers than just "update for fun" (considering angular was labeled like hard to star with... )

what you think ? 

cc/ @alxhub

**Comment by elvisbegovic** [comment]
I checked this : https://downforeveryoneorjustme.com/for-med-large-angular-project-standalone-migration-prune-ngmodule.com

It seems same for everybody ? are we all aware that this is a blocking issue for medium/large to switch project to standalone automatically ? 
(what is med project = >1500comp / >300 lazymodule).

**Comment by ak99372** [comment]
Angular has been shoving standalone components down everyone's throat past 5 versions now it is the default component. Don't bother to mention these limitations nowhere in docs, don't offer alternative, introduce bunch of random magic functions instead and demo how great this works in TODO example, DONE. It's s...t like this devs hate Angular and opt to different frameworks.

**Comment by elvisbegovic** [comment]
Dude don’t dream, migration tool was never a high priority, its better havin cool new todo app without context 100 sloc. 🤷‍♂️
